### PR TITLE
Allow Scaffolder to accept props for custom `TemplateListPage` and `TemplateWizardPage`

### PR DESCRIPTION
**Pull request by minkimcello**
## Motivation 

Replaces #16873.

✅ ~~Should merge #16949 first.~~

## Approach

_✨🚨 Moved `TemplateGroupFilter` and `TemplateGroups` to the `scaffolder-react` package_

The alpha scaffolder plugin will now accept entire pages as props:

```diff
// ./packages/app/src/App.tsx

+ import { MyListPage } from './MyListPage';
+ import { MyWizardPage } from './MyWizardPage';

<Route
  path="/create"
  element={
    <NextScaffolderPage
      groups={[
        { ... },
      ]}
      components={{
+        TemplateListPageComponent: MyListPage,
+        TemplateWizardPageComponent: MyWizardPage,
      }}
    />
  }
>
```

In the screenshots below, I copied both components over to the example app and just updated the header title:

Screenshot 2023-03-16 at 1 33 11 PM

- Some of the modules were not being exported that would allow a user to copy/paste the default list/wizard pages into their own app and replace the local import sources to `@backstage/plugin-scaffolder/alpha` so I went through each import to make sure they are exported
- In order to create your own `TemplateListPage` and `TemplateWizardPage`, users will have to add `@backstage/scaffolder-common` and `@backstage/types` to their app

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] ~~Added or updated documentation~~ _⚠️ The changes proposed are for the alpha version of the scaffolder_
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by jhaals** [comment]
Leaving this one in your capable hands @benjdlambert

File: plugins/scaffolder/alpha-api-report.md [hunk 1]
```diff
@@ -16,6 +20,14 @@ import { ScaffolderTaskOutput } from '@backstage/plugin-scaffolder-react';
 import { SubRouteRef } from '@backstage/core-plugin-api';
 import { TemplateEntityV1beta3 } from '@backstage/plugin-scaffolder-common';
 
+// @alpha
+export const CategoryPicker: () => JSX.Element | null;
```
Comment (benjdlambert) [inline plugins/scaffolder/alpha-api-report.md:24]: Thinking that these components should be moved to `-react` instead and be exported from there. Doesn't make sense for them to live in the plugin anymore.

File: plugins/scaffolder/alpha-api-report.md [hunk 2]
```diff
@@ -16,6 +20,14 @@ import { ScaffolderTaskOutput } from '@backstage/plugin-scaffolder-react';
...
+
+// @alpha (undocumented)
+export function ContextMenu(
```
Comment (benjdlambert) [inline plugins/scaffolder/alpha-api-report.md:27]: `@backstage/plugin-scaffolder-react`

File: plugins/scaffolder/alpha-api-report.md [hunk 3]
```diff
@@ -60,11 +74,64 @@ export const nextSelectedTemplateRouteRef: SubRouteRef<
   PathParams<'/templates/:namespace/:templateName'>
 >;
 
+// @public (undocumented)
+export const registerComponentRouteRef: ExternalRouteRef<undefined, true>;
```
Comment (benjdlambert) [inline plugins/scaffolder/alpha-api-report.md:78]: Don't think that we should export this. We can just leave the authors to implement an explicit `routeRef` if they need one, or if they don't ignore and don't include the component. This was only done to make it configurable to show a button or not depending on if they bound the route ref in the plugin.

File: plugins/scaffolder/alpha-api-report.md [hunk 4]
```diff
@@ -60,11 +74,64 @@ export const nextSelectedTemplateRouteRef: SubRouteRef<
...
+
+// @alpha
+export const RegisterExistingButton: (
```
Comment (benjdlambert) [inline plugins/scaffolder/alpha-api-report.md:81]: I'm also in two minds to export this too - feels pretty small and maybe not worth it?

File: plugins/scaffolder/alpha-api-report.md (hunk 2 above)
Comment (benjdlambert) [reply plugins/scaffolder/alpha-api-report.md:27]: Probably look for a better name here too for where it's used etc.

File: plugins/scaffolder/alpha-api-report.md (hunk 1 above)
Comment (benjdlambert) [reply plugins/scaffolder/alpha-api-report.md:24]: Also rename?

**Review by benjdlambert — CHANGES_REQUESTED** [review]
Hey :wave:

Thanks for this - some comments for a first pass. Basically just moving some bits round and renaming the exports, also some questions about if it's work exporting some things.

File: plugins/scaffolder/alpha-api-report.md (hunk 2 above)
Comment (minkimcello) [reply plugins/scaffolder/alpha-api-report.md:27]: [`ContextMenu` imports three route refs](https://github.com/backstage/backstage/blob/310b781eac12b52a031442b77e12a57ed5bc39b0/plugins/scaffolder/src/next/TemplateListPage/ContextMenu.tsx#L32-L36) from within the same package.

Should I convert the three boolean args into route ref args?

File: plugins/scaffolder/alpha-api-report.md (hunk 1 above)
Comment (minkimcello) [reply plugins/scaffolder/alpha-api-report.md:24]: How's `TemplateCategoryPicker`?

Created a separate PR: #16949

File: plugins/scaffolder/alpha-api-report.md (hunk 2 above)
Comment (benjdlambert) [reply plugins/scaffolder/alpha-api-report.md:25]: Yep, either that, or providing `onClick` handlers for each of them.

Comment (benjdlambert) [reply plugins/scaffolder/alpha-api-report.md:25]: And I think that might be outdated now as those routeRefs don't exist anymore, they're just the same `routeRefs` as of the release tomorrow :)

**Comment by benjdlambert** [comment]
@minkimcello I amended this PR a little bit on top of the previously merged refactor to move out some components. Realised that we should export the `TemplateGroups` component too, rather than it coming from the plugin package itself. So had to move that around and fix the props for that too. Thinking that this is looking pretty good now, and would like to get it in this release today if possible.

**Comment by minkimcello** [comment]
> @minkimcello I amended this PR a little bit on top of the previously merged refactor to move out some components. Realised that we should export the TemplateGroups component too, rather than it coming from the plugin package itself. So had to move that around and fix the props for that too. Thinking that this is looking pretty good now, and would like to get it in this release today if possible.

Updated the PR and rebased. Everything looks good to me too.

**Comment by minkimcello** [comment]
> Thank you for contributing to Backstage! The changes in this pull request will be part of the `1.13.0` release, scheduled for Tue, 18 Apr 2023.

😩

**Comment by benjdlambert** [comment]
Relax @minkimcello it's going to be in the `-next` release today 😝

**Comment by benjdlambert** [comment]
That's the next `stable` release.

**Pull request merged**